AISKU/src/AISku.ts

```diff
@@ -200,8 +200,47 @@ export class AppInsightsSku implements IApplicationInsights {
 
             // Will get recalled if any referenced values are changed
             _addUnloadHook(onConfigChange(cfgHandler, () => {
-                if (_config.connectionString) {
-                    const cs = parseConnectionString(_config.connectionString);
+                let configCs =  _config.connectionString;
+                
+                if (isPromiseLike(configCs)) {
+                    let ikeyPromise = createAsyncPromise<string>((resolve, reject) => {
+                        doAwaitResponse(configCs, (res) => {
+                            let curCs = res.value;
+                            let ikey = _config.instrumentationKey;
+                            if (!res.rejected && curCs) {
+                                // replace cs with resolved values in case of circular promises
+                                _config.connectionString = curCs;
+                                let resolvedCs = parseConnectionString(curCs);
+                                ikey = resolvedCs.instrumentationkey || ikey;
+                            }
+                            resolve(ikey);
+                        });
+
+                    });
+
+                    let urlPromise = createAsyncPromise<string>((resolve, reject) => {
```

Comment:
There are some optimizations we could do here, but this works because promises support having multiple resolve / reject listeners -- but probably not necessary for v1. 

As it's probably messier, where you would extract out the `parseConnectionString()` so it's only called once and as part of this you would cause the iKey and endpoint Promises to get resolved, which means they need to be created before the doAwaitResponse for the connection string and their resolve / reject functions put into external variables that could be called from within the doAwaitResponse...

Reply:
Thanks Nev!  I will do a minifications and optimizations(on other parts as well)